sled-agent/src/storage_manager.rs

```diff
@@ -863,13 +880,177 @@ impl StorageManager {
+    ///
+    /// If errors occur, an arbitrary "one" of them will be returned, but a
+    /// best-effort attempt to add all disks will still be attempted.
+    pub async fn ensure_using_exactly_these_disks<I>(
+        &self,
+        unparsed_disks: I,
+    ) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = crate::hardware::UnparsedDisk>,
+    {
+        let mut new_disks = HashMap::new();
+
+        // We may encounter errors while parsing any of the disks; keep track of
+        // any errors that occur and return any of them if something goes wrong.
+        //
+        // That being said, we should not prevent access to the other disks if
+        // only one failure occurs.
+        let mut err: Option<Error> = None;
+
+        // Ensure all disks conform to the expected partition layout.
+        for disk in unparsed_disks.into_iter() {
+            match crate::hardware::Disk::new(&self.log, disk).map_err(|err| {
+                warn!(self.log, "Could not ensure partitions: {err}");
+                err
+            }) {
+                Ok(disk) => {
+                    let devfs_path = disk.devfs_path().clone();
+                    new_disks.insert(devfs_path, disk);
+                }
+                Err(e) => {
+                    warn!(self.log, "Cannot parse disk: {e}");
+                    err = Some(e.into());
+                }
+            };
+        }
+
+        let mut disks = self.disks.lock().await;
+
+        // Steal the old contents of `disks`; we'll rebuild it below by
+        // inserting an entry for every element of `new_disks`, so replace it
+        // with a hash map that's already appropriately sized. If we return
+        // before completing the loop over `new_disks` below, we will leave
+        // `self.disks` empty or incomplete.
+        let mut old_disks = HashMap::with_capacity(new_disks.len());
+        std::mem::swap(&mut *disks, &mut old_disks);
+
+        // Remove disks that don't appear in the "new_disks" set.
+        //
+        // This also accounts for zpools and notifies Nexus.
+        let keys_to_be_removed = old_disks
+            .iter()
+            .filter(|(key, old_disk)| {
+                // If this disk appears in the "new" and "old" set, it should
+                // only be removed if it has changed.
+                //
+                // This treats a disk changing in an unexpected way as a
+                // "removal and re-insertion".
+                if let Some(new_disk) = new_disks.get(*key) {
+                    // Changed Disk -> Disk should be removed.
+                    new_disk != *old_disk
+                } else {
+                    // Not in the new set -> Disk should be removed.
+                    true
+                }
+            })
+            .map(|(key, _)| key);
+        for key in keys_to_be_removed {
+            if let Err(e) = self.delete_disk_locked(&mut disks, &key).await {
```

Comment:
I don't think this is right - at this point `disks` is empty (because we swapped it on line 940), so `delete_disk_locked` won't do anything. Fixing this seems tricky. Playing with it a bit.

Reply:
My short-term workaround: I'm going back to my old mechanism of updating the set of disks in place.

This desperately needs to be refactored to have some better unit tests without the tight coupling to real hardware